### ISSUE TITLE
Can't write more than 2GB HDF5 datasets in parallel

### ISSUE THREAD

**Issue by rtobar** [ICRAR/VELOCIraptor-STF#88]
When using the inputs and configuration from #87, and after running with the fix for the original issue, VR crashes with the following problem:

```
[0000] [1123.397] [ info] io.cxx:1292 Saving SO particle lists to lala.catalog_SOlist.0
[0001] [1123.397] [ info] io.cxx:1292 Saving SO particle lists to lala.catalog_SOlist.0
[0003] [1123.403] [ info] io.cxx:1292 Saving SO particle lists to lala.catalog_SOlist.0
[0004] [1123.395] [ info] io.cxx:1292 Saving SO particle lists to lala.catalog_SOlist.0
[0005] [1123.395] [ info] io.cxx:1292 Saving SO particle lists to lala.catalog_SOlist.0
[0006] [1123.403] [ info] io.cxx:1292 Saving SO particle lists to lala.catalog_SOlist.0
[0007] [1123.403] [ info] io.cxx:1292 Saving SO particle lists to lala.catalog_SOlist.0
[0002] [1123.409] [ info] io.cxx:1292 Saving SO particle lists to lala.catalog_SOlist.0
HDF5-DIAG: Error detected in HDF5 (1.8.20) MPI-process 2:
  #000: H5Dio.c line 322 in H5Dwrite(): can't prepare for writing data
    major: Dataset
    minor: Write failed
  #001: H5Dio.c line 403 in H5D__pre_write(): can't write data
    major: Dataset
    minor: Write failed
  #002: H5Dio.c line 846 in H5D__write(): can't write data
    major: Dataset
    minor: Write failed
  #003: H5Dmpio.c line 527 in H5D__contig_collective_write(): couldn't finish shared collective MPI-IO
    major: Low-level I/O
    minor: Write failed
  #004: H5Dmpio.c line 1397 in H5D__inter_collective_io(): couldn't finish collective MPI-IO
    major: Low-level I/O
    minor: Can't get value
  #005: H5Dmpio.c line 1441 in H5D__final_collective_io(): optimized write failed
    major: Dataset
    minor: Write failed
  #006: H5Dmpio.c line 295 in H5D__mpio_select_write(): can't finish collective parallel write
    major: Low-level I/O
    minor: Write failed
  #007: H5Fio.c line 169 in H5F_block_write(): write through metadata accumulator failed
    major: Low-level I/O
    minor: Write failed
  #008: H5Faccum.c line 823 in H5F__accum_write(): file write failed
    major: Low-level I/O
    minor: Write failed
  #009: H5FDint.c line 254 in H5FD_write(): addr overflow, addr = 572724, size=18446744071687627232, eoa=12286439660
    major: Invalid arguments to routine
    minor: Address overflowed
HDF5-DIAG: Error detected in HDF5 (1.8.20) MPI-process 7:
  #000: H5Dio.c line 322 in H5Dwrite(): can't prepare for writing data
    major: Dataset
    minor: Write failed
  #001: H5Dio.c line 403 in H5D__pre_write(): can't write data
    major: Dataset
    minor: Write failed
  #002: H5Dio.c line 846 in H5D__write(): can't write data
    major: Dataset
    minor: Write failed
  #003: H5Dmpio.c line 527 in H5D__contig_collective_write(): couldn't finish shared collective MPI-IO
    major: Low-level I/O
    minor: Write failed
  #004: H5Dmpio.c line 1397 in H5D__inter_collective_io(): couldn't finish collective MPI-IO
    major: Low-level I/O
    minor: Can't get value
  #005: H5Dmpio.c line 1441 in H5D__final_collective_io(): optimized write failed
    major: Dataset
    minor: Write failed
  #006: H5Dmpio.c line 295 in H5D__mpio_select_write(): can't finish collective parallel write
    major: Low-level I/O
    minor: Write failed
  #007: H5Fio.c line 169 in H5F_block_write(): write through metadata accumulator failed
    major: Low-level I/O
    minor: Write failed
  #008: H5Faccum.c line 823 in H5F__accum_write(): file write failed
    major: Low-level I/O
    minor: Write failed
  #009: H5FDint.c line 254 in H5FD_write(): addr overflow, addr = 572724, size=18446744072925879424, eoa=12286439660
    major: Invalid arguments to routine
    minor: Address overflowed
Failed to write dataset: Particle_IDsFailed to write dataset: Particle_IDs

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 2
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 7
```

**Comment by MatthieuSchaller** [comment]
Adding something in case it's useful. My tests did not crash but were just hanging.

MPI + OpenMP. All on a single node (i.e. no fabric involved) in the same setup as #87 and with hdf5 1.10.6

**Comment by rtobar** [comment]
Thanks @MatthieuSchaller for the details, that's useful to know. While trying to reproduce locally (with a smaller dataset) I think I hit more-or-less the same problem detected by valgrind, but not leading to a crash/hangup:

```
==13495== Syscall param pwritev(vector[0]) points to uninitialised byte(s)
==13495==    at 0x56E25A8: pwritev64 (pwritev64.c:26)
==13495==    by 0x56E25A8: pwritev (pwritev64.c:24)
==13495==    by 0xA2C6E25: mca_fbtl_posix_pwritev (in /usr/lib/x86_64-linux-gnu/openmpi/lib/openmpi3/mca_fbtl_posix.so)
==13495==    by 0x9A17514: ??? (in /usr/lib/x86_64-linux-gnu/openmpi/lib/openmpi3/mca_fcoll_vulcan.so)
==13495==    by 0x9A18883: mca_fcoll_vulcan_file_write_all (in /usr/lib/x86_64-linux-gnu/openmpi/lib/openmpi3/mca_fcoll_vulcan.so)
==13495==    by 0x9A34DBF: mca_common_ompio_file_write_at_all (in /usr/lib/x86_64-linux-gnu/libmca_common_ompio.so.41.19.3)
==13495==    by 0x99EA24A: mca_io_ompio_file_write_at_all (in /usr/lib/x86_64-linux-gnu/openmpi/lib/openmpi3/mca_io_ompio.so)
==13495==    by 0x4DDDD7F: PMPI_File_write_at_all (in /usr/lib/x86_64-linux-gnu/libmpi.so.40.20.3)
==13495==    by 0x514892A: ??? (in /usr/lib/x86_64-linux-gnu/libhdf5_openmpi.so.103.2.0)
==13495==    by 0x4F971FD: H5FD_write (in /usr/lib/x86_64-linux-gnu/libhdf5_openmpi.so.103.2.0)
==13495==    by 0x4F76302: H5F__accum_write (in /usr/lib/x86_64-linux-gnu/libhdf5_openmpi.so.103.2.0)
==13495==    by 0x5072658: H5PB_write (in /usr/lib/x86_64-linux-gnu/libhdf5_openmpi.so.103.2.0)
==13495==    by 0x4F8083B: H5F_block_write (in /usr/lib/x86_64-linux-gnu/libhdf5_openmpi.so.103.2.0)
==13495==  Address 0x108e3238 is 504 bytes inside a block of size 16,777,216 alloc'd
==13495==    at 0x483C7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==13495==    by 0x9A18483: mca_fcoll_vulcan_file_write_all (in /usr/lib/x86_64-linux-gnu/openmpi/lib/openmpi3/mca_fcoll_vulcan.so)
==13495==    by 0x9A34DBF: mca_common_ompio_file_write_at_all (in /usr/lib/x86_64-linux-gnu/libmca_common_ompio.so.41.19.3)
==13495==    by 0x99EA24A: mca_io_ompio_file_write_at_all (in /usr/lib/x86_64-linux-gnu/openmpi/lib/openmpi3/mca_io_ompio.so)
==13495==    by 0x4DDDD7F: PMPI_File_write_at_all (in /usr/lib/x86_64-linux-gnu/libmpi.so.40.20.3)
==13495==    by 0x514892A: ??? (in /usr/lib/x86_64-linux-gnu/libhdf5_openmpi.so.103.2.0)
==13495==    by 0x4F971FD: H5FD_write (in /usr/lib/x86_64-linux-gnu/libhdf5_openmpi.so.103.2.0)
==13495==    by 0x4F76302: H5F__accum_write (in /usr/lib/x86_64-linux-gnu/libhdf5_openmpi.so.103.2.0)
==13495==    by 0x5072658: H5PB_write (in /usr/lib/x86_64-linux-gnu/libhdf5_openmpi.so.103.2.0)
==13495==    by 0x4F8083B: H5F_block_write (in /usr/lib/x86_64-linux-gnu/libhdf5_openmpi.so.103.2.0)
==13495==    by 0x5146A70: H5D__mpio_select_write (in /usr/lib/x86_64-linux-gnu/libhdf5_openmpi.so.103.2.0)
==13495==    by 0x513EAD5: ??? (in /usr/lib/x86_64-linux-gnu/libhdf5_openmpi.so.103.2.0)
```

Probably the final effect of the underlying problem is different on different platforms/compilers/library combinations. I'll post again once I have a better idea of what's going on.

**Comment by MatthieuSchaller** [comment]
Yes, that's plausible. Our hdf5 is built on top of Intel-MPI so will possibly crash in a different way in the depth of the communcation stack. or hang.

**Comment by rtobar** [comment]
As an update, I've managed to reduce this problem to a smaller reproducible example added in 0a37d4d (mind you the build is currently broken, but that will be taken care of later on). The test sets up local arrays of sizes similar to those causing the problem, and then tries to write them into a single HDF5 Dataset in parallel.

The behavior of the test is a weird one. Under small loads (say, if the array sizes are 1/10th of the original) everything goes well. As they increase the test fails, but only after running for a while instead of immediately.

The test doesn't use the hdf5 routines directly but the `H5OutputFile` class that is used by the rest of VR, plus some extra MPI setup that is required to use the class correctly. I'm a bit torn between thinking whether this is a parallel HDF5 problem or a VR problem, but these are not huge data sizes and HDF5 is a heavily battle-tested library, so I'd assume it's the latter.

I'll try to strip the test code from VR-specifics little by little to see if I can converge into a working test, potentially pinpointing what's wrong on the VR side, if anything.

**Comment by MatthieuSchaller** [comment]
Ah, one thing come to mind. parallel hdf5 cannot write more than 2GB / MPI rank at one time in a single call. This is because of the limitations of MPI-IO used at a lower level. If you have more than 20GB (say) on 10 nodes then you will need to write 2 GB / node and then write the rest. 

I found that out the hard way a while back with SWIFT. Though I don't recall what happens if you go > 2GB. Does it crash? or hang? don't remember.

**Comment by rtobar** [comment]
Yes, I've just been educating myself on this very same limitation, I didn't know about it! I've been playing a bit more with the test code I committed above and at this point I can't really blame VR for anything obvious, so it really seems to be a limitation on HDF5 and MPI libraries (or combination thereof).

What I could gather so far:
 * The limitation was allegedly [fixed in 1.10.2](https://www.hdfgroup.org/2018/04/why-should-i-care-about-the-hdf5-1-10-2-release/). The problem on this issue #88 was originally reported against 1.8.20, so in short we hit an unsupported case, and we'll either have to restrict the HDF5 versions we build against, or indeed for a write-in-chunks logic. @MatthieuSchaller did you end up doing that indeed in swift, or did you simply constrain the HDF5 versions you support?
 * With hdf5 1.10.6 in cosma we see a different problem: it hangs instead of reporting an error and crashing. I can't find anything obvious in the 1.10.7 or 1.12.0 release notes indicating this is a problem that has been fixed, so I wonder whether it's again a situation that we can handle in our code by just avoiding writes of >= 2^31 bytes.
 * Parallel, compressed writing seems to be limited to MPI 3, but that should be OK in all our cases I think.
 * Locally I'm testing with hdf5 1.10.6 with OpenMPI 4.0.3, so in theory I should be free from the limitation. Apparently in the past there were build/runtime issues with OpenMPI, but not anymore [since 1.10.5](https://portal.hdfgroup.org/display/knowledge/OpenMPI+Build+Issues) and for certain minimum OpenMPI versions (which mine is part of). I can indeed write with multiple ranks that contribute more than 2GB of data, but...
 * Locally I'm experiencing other types of problems, which might or might not be reproducible in cosma. For example, I can write a dataset with one rank contributing 2^31 bytes and another contributing 8, but I can't do the same with a single rank writing 2^31 bytes! It looks a bit like this: https://github.com/HDFGroup/hdf5/issues/549
 * There have been a number of bug fixes throughout the different HDF5 1.10 releases around parallel IO, so this seems to be a bit of a minefield at the moment.

I guess the best solution is to bite the bullet and write data in < 2GB chunks when doing parallel I/O.

**Comment by MatthieuSchaller** [comment]
I decided to work around the limitation by writing in per-rank chunks of 2GB. If it is useful, you can check SWIFT's parallel_io.c file.
The 2GB limitation is also good in principle as it means, we don't actually allocate very large buffer arrays. And neither does MPI behind our back.

It could be that the crash/hang behaviour is dependent on the library version and underlying mpi implementation. 

Note that so far I have not been brave enough to swittch on compression when using parallel-hdf5 in production. It's also not super fast in MPI+X mode as only one core (the one with the MPI object) does the compression.

**Comment by MatthieuSchaller** [comment]
FYI, if I test SWIFT with hdf5 1.10.6 and try to write >2GB / rank I hang. 
That is using Intel-MPI over the fabric.

So I do not believe the version 1.10.2 actually fixed all of this. 

I had never seen that part of their release notes and assumed that things were stuck in this state. Good thing I didn't see it and somehow kept the cunking version.
Anyway, when I get time to do it, I'll submit a bug report to hdf5 once more. 

As for VR, I think the only way out is to write in chunks or not write such large arrays in parallel. Painful but looks like the only realistic short-term solution.

**Comment by rtobar** [comment]
Yes, the hdf5 status is clearly a bit flaky. Good to know (in a way) that SWIFT is also affected by this, it confirms this is an hdf5-level issue.

I wanted to do the write-per-chunk changes yesterday, but while working on this issue I've given a deeper dive into the HDF5 writing code in VR. Since then I've taken the opportunity to clean things up a bit, remove code duplication, add test code, etc, so I was busy finishing that. Today I'll hopefully be able to finally implement the chunked writes.

**Comment by rtobar** [comment]
I think I got this fixed now, finally, in the `issue-88` branch. I'm currently running tests on cosma, and with my test program I can write from 8 ranks with the original sizes and data type that triggered the error. When running VR to reproduce the original error I hit a different error, but I think it's due to not enough memory being available -- but that was running with 1 node. I'm re-running now with 2 nodes and 8 ranks, and hopefully that will show the error is gone.

After that I'll still need to do some cleanups (e.g., builds are failing in Travis, I put too many new logs, etc). Once that's done I can hand over this for further testing before merging to the master branch.

Also, to avoid issues like the one I faced now (i.e., running out of memory) I'll work on adding a new option to specify the writing chunk size, but will still default to 2^31-1 (and this also acts as the limit). That can be handled separately though.

**Comment by rtobar** [comment]
OK, after finding another issue (see #100), and fixing it, I was now able to run VR with 8 ranks in two nodes, and finally confirm the fix is working!

Like I mentioned earlier I will clean up the code a bit before merging. I'll put an update here.

**Comment by rtobar** [comment]
I just tried this again on `cosma6` and the writing of the SO catalog seems indeed to be working as expected. I was waiting to test this together with the changes in `issue-71` to (partially) address #71 and be able to run the experiment on one node instead of two, but even when running in one node I'm still running into #102. That's a different issue though, so I'm fairly happy with what's in `issue-88` so that it can go into the `master` branch.

@MatthieuSchaller I'd feel better if you or someone else tested these changes first before merging. Be aware of the further problems though, specially #102, we'll tackle that separately.

**Comment by MatthieuSchaller** [comment]
Happy to do that. Any setup in particular that would help you?

**Comment by rtobar** [comment]
Probably just make sure the original use case now works (bar #102) and that the generated SO output makes sense.

**Comment by MatthieuSchaller** [comment]
I have run `issue-88` on a large SWIFT-EAGLE setup with parallel-hdf5 switched on with 192 MPI ranks. 
The code did hang while writing the `.properties` file. Note that I don't think any rank is writing more than 2GB. 

The writing of `.SO_list` (2TB in total) worked fine. It was very slow, however. The writing itself took 2.5 hours. I had not used Lustre striping so that may be part of the problem.

**Comment by rtobar** [comment]
> The writing itself took 2.5 hours

Ouch! Yes, Lustre striping might improve things I guess, but any imbalance in the work distribution across ranks (which I don't really know how it works in VR) will also contribute to slow writes I guess. Locally in my laptop I get 1-rank executions and 4-rank executions both at speeds that go up to ~300 MB/s. That seems too fast for my SSD, so there's probably some funny fs-level caching at play.

A bit of a side question: has VR parallel HDF5 writing been used in anger before? I'm a bit puzzled that so many issues were found around this subject (first #87, then this ticket, then #100 and also #102), but maybe it's a particular combination of inputs/configs that hadn't occurred before.

**Comment by MatthieuSchaller** [comment]
> Ouch! Yes, Lustre striping might improve things I guess, but any imbalance in the work distribution across ranks (which I don't really know how it works in VR) will also contribute to slow writes I guess. Locally in my laptop I get 1-rank executions and 4-rank executions both at speeds that go up to ~300 MB/s. That seems too fast for my SSD, so there's probably some funny fs-level caching at play.

A similar run without parallel-hdf5 on the snapshot just before showed a huge imbalance on the writing time. Some ranks finished in <2min and some took 40min. So I do not think it's entirely a problem of parallel writing. I fear that VR is hugely imbalanced in certain scenarios. That is also the case when monitoring the memory usage. Some ranks can use ~10x the memory of others.
The pure DMO cases might do better however.

> A bit of a side question: has VR parallel HDF5 writing been used in anger before? I'm a bit puzzled that so many issues were found around this subject (first #87, then this ticket, then #100 and also #102), but maybe it's a particular combination of inputs/configs that hadn't occurred before.

I haven't used it much myself until recently. Most of our runs were small enough that they either fit on one node and OpenMP was enough or they ran on a few number of nodes and having a handful of output files per output was fine.
Now, when going to 100s of nodes, it gets a bit more tricky. 

I think it was added relatively recently. Have you guys used it when running on GADI? 
It could also be indeed that we see problems with our full-physics configs. Some issues may not have been apparent in the DMO runs.

**Comment by rtobar** [comment]
Thanks @MatthieuSchaller for the further information. I'm not sure what @doctorcbpower has been using in Gadi, but as you say these issues were probably triggered by the full-physics + extra gas/star/bh quantities configs.

Anyhow, now that #102 is finally solved I feel much more confident about merging these changes. I'm doing a final run on cosma to verify all issues are gone, then I'll create a PR to trigger the builds and execution in Travis in case there are any issues those executions could spot, and finally will merge `issue-88` back to the `master`.

**Comment by doctorcbpower** [comment]
For the large N runs we have been doing, VR writes distributed HDF rather than a single HDF. It's also pure, single species, N-body, so that might also make a difference.

**Comment by rtobar** [comment]
Done, changes have been merged now to the `master` branch via #111.

**Comment by rtobar** [comment]
And thanks @doctorcbpower for the info, I think that also clarifies why we hadn't stumbled into these issues in Gadi.

**Comment by MatthieuSchaller** [comment]
One other possible reason why this hasn't showed up before is that the SO lists seem to be randomly "wrong". 
The lists are typically too large and sometimes much much too large, with many more particles listed than were in the snapshot. 

That seems to be more often the case when running larger volumes and possibly also more often with hydro but I have little stats. 

The list uses a conservative estimate for what the SO radius will be and outputs all particles in there. This radius seems to be > R_200 but is a simple estimate. In some cases, we see it be >>> R_200 leading to lists which are much too big. And then possibly to this problem. 
It's a story for another thread (see Bert's message on slack) but a possible additional explanation for why this wasn't seen.

**Comment by rtobar** [comment]
Thanks @MatthieuSchaller for the further insights. Yes, I had seen Bert's message in Slack, but I'm not really familiar with the technicalities around the SO logic, so I can't comment on that.

Your mention of "randomly" wrong though reminds me of #90, where uninitialised values are seen all over the place, and seem to affect the SO code. I did start some time ago to clean that up similarly to how I did with the `Option` class recently (8f380fc), and in the process take the opportunity to remove unused fields, etc. I didn't get the chance to finish that work, but it's already quite advanced, so I'll upload what I have so far.

**Comment by doctorcbpower** [comment]
This was a serious problem in the DMO mode but it was fixed by Rhys - when the enclosed density and radius was being interpolated to get exactly e.g. 200 times critical, it could shoot off to some crazily large value . I am surprised that it is an issue in the hydro. I can check to see if something odd happens if hydro is switched on.

**Comment by MatthieuSchaller** [comment]
Ah, let me clarify. We don't (yet?) think it's random in that sense. It's more that the estimate use for the radius of the SO list is based on quantities which may not be the ideal choice and can push up the value for "physical" reasons.

@doctorcbpower the R_xxx and friends are all correct. However, (surprisingly?), the list of particles for the SO list does not use that. It uses a much earlier guesstimate made in the code. Something close to an upper-limit one would use to indeed correctly bissect and find R_xxx. But, it then persists with this original estimate and writes out all the particles in there. 
And in some rare cases, this is very very large.

**Comment by rtobar** [comment]
@MatthieuSchaller thanks for the clarification, understood now ;)

When you have a chance, would you be able to open an issue to track this separately? Otherwise we'll keep losing these bits of important information in the jungle of 2GB hdf5 output chunking.

**Comment by MatthieuSchaller** [comment]
Yes, will do. Once we have slightly more info.

**Comment by doctorcbpower** [comment]
@MatthieuSchaller - sigh - why doesn't this surprise me?!
